tests/Doctrine/Tests/ORM/Functional/Ticket/GH7941Test.php

```diff
@@ -64,6 +65,18 @@ public function typesShouldBeConvertedForDQLFunctions(): void
         foreach ($query->getResult() as $i => $item) {
             $product = self::PRODUCTS[$i];
 
+            // !!!!! TEMPORARY HACK
+            // Just to make sure this is the only remaining issue to solve as I'm not sure if the issue here is the
+            // testcase or the inconsistent return type.
+            // mysqli returns the `square_root` column as a float but interestingly it doesn't return the average
+            // as a float...?
+            // All other drivers return the sqrt value as a string
+            if ($this->getEntityManager()->getConnection()->getDriver() instanceof MysqliDriver) {
+                $this->assertIsFloat($item['square_root']);
+                $this->addWarning('Manually casting inconsistent `float` value to string for test to pass');
+                $item['square_root'] = (string) $item['square_root'];
```

Comment:
This test was failing with the mysqli driver because assertRegExp now strictly requires a string value.

All other drivers are returning a string for the `square_root` column, but mysqli is returning a float. Oddly this is the only affected column, the others (including the result of AVG in the query above) are all strings.

The test was introduced in #7941 and I can't spot anything since that would be causing it to fail. My hunch is it was always getting a float on mysqli but it wasn't a problem due to the soft typing in older phpunit.

I've had a bit of a dig in internals and as far as I can see:

* SqlWalker::walkSelectExpression is happily calling ResultSetMapping::addScalarResult with a `string` type for all the columns in these queries
* AbstractHydrator::gatherRowData is correctly getting a `Doctrine\DBAL\Types\StringType` type object for the square_root column
* `StringType::convertToPhpValue` falls back to the parent `Type::convertToPhpValue` which just returns the incoming value
* With the mysqli extension, `->convertToPhpValue()` is called with a `double`, but with pdo_mysql (and all other drivers) it's called with a `string`.

So unless I'm missing something I think it's just that the `mysqli` extension returns a `double` for columns that call the `SQRT` function.

If that's right then there are two options:

* If the return type should be guaranteed to be a string across all drivers then I guess `StringType->convertToPhpValue` needs to explicitly cast the return value - but that feels like quite a major change that could cause problems elsewhere.
* If it's expected that the return type is inconsistent / undefined then I can just tidy up the fix in the test so that it will accept either float or string. Instead of regexp, I'd be tempted to compare them using `->assertEqualsWithDelta()` to match numerically with a reasonable level of precision. That would probably be more future proof, especially if the future plan is to manually cast these functions to float as discussed in #7941.

How should I proceed?

Reply:
Since types are defined in the DBAL (so another package) I think it will be easier to go with option 2 (and `->assertEqualsWithDelta()`. @morozov, are you maybe aware of this discrepancy of returned type between extensions?

Reply:
@greg0ire I believe it's a common thing for many of the DB drivers in PHP. IIRC, `mysqli` may [cast](https://www.php.net/manual/en/mysqli.insert-id.php) an integer to a string if there's a risk of it causing integer overflow.

Reply:
Thanks @morozov OK, makes sense - I'll update the test accordingly.

Reply:
Done in https://github.com/doctrine/orm/pull/8532/commits/62bf9b83ca59d6552eedbc3582290a8151702eca (to be squashed up). Turns out the actual returned values vary quite a bit, even in the level of precision in the drivers that return strings. I was surprised mysql only gives 4 d.p. for the average...